homeassistant/components/thethingsnetwork/__init__.py

```diff
@@ -33,10 +27,65 @@
 
 async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
     """Initialize of The Things Network component."""
-    conf = config[DOMAIN]
-    app_id = conf.get(CONF_APP_ID)
-    access_key = conf.get(CONF_ACCESS_KEY)
 
-    hass.data[DATA_TTN] = {TTN_ACCESS_KEY: access_key, TTN_APP_ID: app_id}
+    if DOMAIN in config:
+        ir.async_create_issue(
+            hass,
+            DOMAIN,
+            "manual_migration",
+            breaks_in_ha_version="2021.10.0",
+            is_fixable=False,
+            severity=ir.IssueSeverity.ERROR,
+            translation_key="manual_migration",
+            translation_placeholders={"domain": DOMAIN},
+        )
 
     return True
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Establish connection with The Things Network."""
+
+    _LOGGER.debug(
+        "Set up %s at %s",
+        entry.data[CONF_APP_ID],
+        entry.data.get(CONF_HOSTNAME, TTN_API_HOSTNAME),
+    )
+
+    # Create coordinator to fetch TTN updates
+    entry.coordinator = coordinator = TTNCoordinator(hass, entry)
```

Comment:
Instead of assigning the coordinator to the entry, please assign it to `hass.data[DOMAIN]`. If you support multiple instances of the integration, please use `hass.data.setdefault(DOMAIN, {})[entry.entry_id]` to store the coordinator

Reply:
yes, I support multiple instances. Most users will likely only have one but each TheThingsNetwork application would need a different instance.

I will use `hass.data.setdefault(DOMAIN, {})` . I would also need to use this for other information saved at runtime in `entrySettings.py`.